source/material_model/rheology/strain_dependent.cc

```diff
@@ -329,16 +226,20 @@ namespace aspect
         // nonlinear solver scheme does a single advection iteration. More than one nonlinear advection
         // iteration will result in the incorrect value of strain being used in the material model, as
         // the compositional fields representing strain are updated through the reaction terms.
-        if (use_strain_weakening)
+        if (weakening_mechanism != none)
           {
             AssertThrow((this->get_parameters().nonlinear_solver ==
                          Parameters<dim>::NonlinearSolver::single_Advection_single_Stokes
                          ||
                          this->get_parameters().nonlinear_solver ==
-                         Parameters<dim>::NonlinearSolver::single_Advection_iterated_Stokes),
+                         Parameters<dim>::NonlinearSolver::single_Advection_iterated_Stokes
+                         ||
+                         this->get_parameters().nonlinear_solver ==
+                         Parameters<dim>::NonlinearSolver::single_Advection_iterated_Newton_Stokes),
```

Comment:
This is a change in functionality. I guess it makes sense to allow this. But have we thought about how this works? As far as I can see, the Newton solver does not compute derivatives with respect to composition, but in a case with strain weakening, the viscosity could depend strongly on the different strain fields, correct?
Using a Newton solver without these terms probably still works much better than using the standard iterated Stokes, but do we think this is something that would be useful to add at some point in the future? Should we note that somewhere?

Reply:
Yes, good point! At minimum, I think we should add a comment noting this. @MFraters - Would this be difficult to implement?

Reply:
Sorry for the late reply, I still have to find a way where github messages where I am mentioned will go to my main mail box :)

The Newton solver currently only uses the viscosity derivatives wrt the strain-rate and the pressure. In the visco-plastic plugin it does this for every composition and then averages the viscosities derivatives for the composition in the same way the viscosities for the compositions are averages. I haven't thought about what would happen with a composition which contains the strain, but by guess would be that it should be alright, because like a normal composition it influences the viscosity.

This means that the Newton solver currently only gives possible quadratic convergence for the nonlinearities from strain-rate and pressure dependence in viscosity. The rest of the nonlinearies have linear/Picard like convergence. 

If the viscosity is really strongly dependent on strain, and you want near quadratic convergence, we would need to compute the derivatives (this can be very easy if you just do a finite difference) and I think we would also need to solve it as a coupled system. Putting everything in a coupled system (we would then probably also put temperture and the other compositions in there) would be probably quite a bit of work because it changes the properties of the matrix we are solving and the way we need to precondition it. @bangerth might have a better idea on how difficult this would be and how that should be done.

relevant lines in visco-plastic:
https://github.com/geodynamics/aspect/blob/f457e03d3d805ae5022adc1b15063053566522e2/source/material_model/visco_plastic.cc#L304-L382